![USPS](https://special.usps.com/static/media/home-test-hero.2099b1df64ebd24bb68a.jpg)

## Who Ordered Free At-Home COVID-19 Tests?

### by Jae Downing

I wanted to know whether there were differences in the characteristics of homes that ordered the free COVID-19 tests compared to those who did not. My hypothesis was that households who ordered the tests would reside in neighborhoods with lower rates of:

* Poverty
* Racial minorities
* Renters

My approach was:

1. Gather a random set of residential addresses
2. [Order COVID-19 tests](https://special.usps.com/testkits) for the households and track whether the household had already ordered the test
3. Geocode Addresses
4. Append Block Groups to Lat and Long & Add in relevant American Community Data
5. Test whether there were differences in block group characteristics

### Step 1: Gather Addresses

I couldn't find a national list of residential addresses. I live in Portland, so I thought I would use a local source. Since an API wasn't available for the residental address files, I first downloaded [Active Addresses](https://gis-pdx.opendata.arcgis.com/datasets/active-address-points/explore?location=45.339300%2C-122.536950%2C9.93) from [Portland Maps](https://www.portlandmaps.com/).  

In [3]:
import pandas as pd
data = pd.read_csv("/Users/downingj/Desktop/Active_Address_Points.csv")
len(data)
#data.head()

/var/folders/y8/np38j_yx7f19x94jw97cg34ss2prd2/T/ipykernel_5729/2993792815.py:2: DtypeWarning: Columns (3,18,23,28,33,34,36,38,44) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/Users/downingj/Desktop/Active_Address_Points.csv")


832735

In [4]:
# Keep only the addresses that are confirmed deliverable
data = data[(data['MAILING_VALIDITY'] == "Confirmed Deliverable")]
len(data)

300669

In [6]:
# Extract a random sample of 1000 addresses after sorting by zipcode
data.sort_values(by=['ZIP_CODE'])
sample = data.sample(n=1000, random_state=1)
sample = pd.DataFrame(sample)
len(sample)

1000

In [7]:
# Create a new csv file with the fields needed to fill out the form

# Since Name is a Required Field, we will call everyone Oregon Resident
sample['firstName'] = "Oregon"
sample['lastName'] = "Resident"
sample['email'] = " "
sample['streetaddress'] = sample['ADDRESS_FULL']
sample['city'] = sample['MAIL_CITY']
sample['state'] = "OR"
sample['zipcode'] = sample['ZIP_CODE']

In [8]:
# Keep fields needed

final = sample[['firstName', 'lastName', 'email', 'streetaddress', 'city', 'state', 'zipcode', 'X', 'Y']]
fn = 'address_thousand.csv'
final.to_csv(fn)

### Step 2: Order the COVID-19 tests & Track Results

I used [selenium](https://www.selenium.dev/) to automate the web browser. I used the tool to act like a human would - going to the website, typing in the information, and ordering the tests.

The code also allows us to make a csv to track three outcomes:
1. Test is successfully ordered.
2. At-home COVID-19 tests have already been ordered for this address. 
3. We couldn't validate your address as complete.

In [ ]:
# Import Packages
from selenium import webdriver
from webdriver_manager.chrome import ChromeDriverManager as CM
from selenium.webdriver.common.by import By
from selenium.webdriver.support.select import Select
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains
from selenium.common.exceptions import NoSuchElementException
from random import randint
from time import sleep
import csv

In [ ]:
def __random_sleep__(minimum=2, maximum=7):
    t = randint(minimum, maximum)
    sleep(t)

In [ ]:
def __scrolldown__(driver):
    ActionChains(driver).send_keys(Keys.PAGE_DOWN).perform()

In [ ]:
def __scrollup__(driver):
    driver.execute_script(
    "window.scrollTo(0, 0);")

In [ ]:
def is_element_present(how, what):
    """Check if an element is present"""
    try:
        driver.find_element(by=how, value=what)
    except NoSuchElementException:
        return False
    return True

In [ ]:
def setupChrome():
    options = webdriver.ChromeOptions()
    driver = webdriver.Chrome(
    executable_path=CM().install(), options=options)
    
    return driver

In [ ]:
def readFromCsv(file):
    rows = []
  
    # reading csv file
    with open(file, 'r') as csvfile:
        # creating a csv reader object
        csvreader = csv.reader(csvfile)
        
        # extracting field names through first row
        fields = next(csvreader)
    
        # extracting each data row one by one
        for row in csvreader:
            rows.append(row)
  
        # get total number of rows
        print("Total no. of rows: "+str(len(rows)))


    return rows

In [ ]:
def fillForm(len,driver, rows):
    for i in range(len+1):
        address_track = []

        driver.get('https://special.usps.com/testkits')

        __scrollup__(driver)

        __random_sleep__(3,5)

        firstName = driver.find_element(By.NAME, "firstName").send_keys(rows[i][0])
        __random_sleep__(1,2)

        lastName = driver.find_element(By.NAME, "lastName").send_keys(rows[i][1])

        #__random_sleep__(1,2)

        #email = driver.find_element(By.NAME, "email").send_keys(rows[i][2])
        #__random_sleep__(1,2)

        #__scrolldown__(driver)
        __random_sleep__(2,3)

        checkBox = driver.find_element(By.ID, "recipientSameAsSender").click()
        __random_sleep__(1,2)

        address = driver.find_element(By.NAME, "address1").send_keys(rows[i][3])
        __random_sleep__(1,2)

        city = driver.find_element(By.NAME, "city").send_keys(rows[i][4])
        __random_sleep__(1,2)

        state = driver.find_element(By.ID, "state")
        stateDD = Select(state)
        stateDD.select_by_value(rows[i][5])
        __random_sleep__(1,2)

        zipCode = driver.find_element(By.NAME, "zipCode").send_keys(rows[i][6])
        __random_sleep__(2,3)

        checkOut = driver.find_element(By.XPATH, "//button[text()= 'Check Out Now']").click()
        __random_sleep__(2,3)

        orderButton = driver.find_element(By.XPATH, "//button[text() = 'Place My Order']").click()
        __random_sleep__(7,10)
        
        order_failed = "/html/body/div[3]/div/div/div[2]/div/div/p"
        order_successfull = "/html/body/div[3]/div/div/div/div/div/div/h1"

        if is_element_present(By.XPATH, order_failed):

            result = driver.find_element(By.XPATH, order_failed)

        elif is_element_present(By.XPATH, order_successfull):
            result = driver.find_element(By.XPATH, order_successfull)

        
        address_track.append(rows[i][3])
        address_track.append(rows[i][6])
        address_track.append(result.text)

        with open("address_track.csv", 'a') as csvfile: 
            csvwriter = csv.writer(csvfile)
            csvwriter.writerow(address_track)

            print(address_track)
            csvfile.close()


        print("Entered Row number: "+str(i+1))

In [ ]:
filename="/Users/downingj/Desktop/address_thousand.csv"

rows = readFromCsv(filename)

driver = setupChrome()

fillForm(len(rows),driver,rows)

### Step 3: Geocode Using Google API

Next, we have to geocode the addresses. I choose to use Google API. 

In [9]:
# Create a single address field
final = pd.read_csv("/Users/downingj/Desktop/Active_Address_Points.csv")
final = final.applymap(str)

In [10]:
final['address']  = final[['streetaddress', 'city', 'state', 'zipcode']].agg(', '.join, axis=1)

In [11]:
final.head()

firstName  lastName email         streetaddress      city state  \
7090      Oregon  Resident          5639 SE LAMBERT ST  Portland    OR   
265966    Oregon  Resident           9826 SE RAMONA ST  Portland    OR   
62469     Oregon  Resident        501 N GRAHAM ST #335  Portland    OR   
216807    Oregon  Resident             641 SE 131ST PL  Portland    OR   
523156    Oregon  Resident         3577 SE MALL ST #40  Portland    OR   

       zipcode            X           Y  \
7090     97206  7662126.775  663555.766   
265966   97266  7673247.742  667977.267   
62469    97227   7645926.81  691662.683   
216807   97233  7682280.435  681342.838   
523156   97202   7656846.84  672324.268   

                                          address  
7090      5639 SE LAMBERT ST, Portland, OR, 97206  
265966     9826 SE RAMONA ST, Portland, OR, 97266  
62469   501 N GRAHAM ST #335, Portland, OR, 97227  
216807       641 SE 131ST PL, Portland, OR, 97233  
523156   3577 SE MALL ST #40, Portland, OR, 97202

In [90]:
import googlemaps
gmaps_key = googlemaps.Client(key="<YOUR KEY HERE>")

In [91]:
from geopy.geocoders import GoogleV3
import geopy.distance
import googlemaps
import numpy as np
import geopandas as gpd
import urllib

In [92]:
#API = 'YOUR KEY HERE'

In [95]:
#Geocode all the addresses and creates rows with
def my_geocoder(row):
    try:
        point = geolocator.geocode(row).point
        return pd.Series({'Latitude': point.latitude, 'Longitude': point.longitude})
    except:
        return None

final[['Latitude', 'Longitude']] = final.apply(lambda x: my_geocoder(x['address']), axis=1)

print("{}% of addresses were geocoded!".format(
    (1 - sum(np.isnan(final["Latitude"])) / len(final)) * 100))

# Drop addresses that were not successfully geocoded
final = final.loc[~np.isnan(final["Latitude"])]
final = gpd.GeoDataFrame(
    final, geometry=gpd.points_from_xy(final.Longitude, final.Latitude))
final.head()

100.0% of addresses were geocoded!


firstName  lastName email         streetaddress      city state  \
7090      Oregon  Resident          5639 SE LAMBERT ST  Portland    OR   
265966    Oregon  Resident           9826 SE RAMONA ST  Portland    OR   
62469     Oregon  Resident        501 N GRAHAM ST #335  Portland    OR   
216807    Oregon  Resident             641 SE 131ST PL  Portland    OR   
523156    Oregon  Resident         3577 SE MALL ST #40  Portland    OR   

       zipcode            X           Y  \
7090     97206  7662126.775  663555.766   
265966   97266  7673247.742  667977.267   
62469    97227   7645926.81  691662.683   
216807   97233  7682280.435  681342.838   
523156   97202   7656846.84  672324.268   

                                          address   Latitude   Longitude  \
7090      5639 SE LAMBERT ST, Portland, OR, 97206  45.467446 -122.604997   
265966     9826 SE RAMONA ST, Portland, OR, 97266  45.480414 -122.562097   
62469   501 N GRAHAM ST #335, Portland, OR, 97227  45.543346 -122.671081   
216807       641 SE 131ST PL, Portland, OR, 97233  45.517647 -122.528236   
523156   3577 SE MALL ST #40, Portland, OR, 97202  45.491123 -122.626488   

                           geometry  
7090    POINT (-122.60500 45.46745)  
265966  POINT (-122.56210 45.48041)  
62469   POINT (-122.67108 45.54335)  
216807  POINT (-122.52824 45.51765)  
523156  POINT (-122.62649 45.49112)

### Step 4: Append Census Blocks and Data

Ok, that was helpful, but this doesn't contain FIPS. R has some great packages for working with Census Data, so let's demonstrate how to switch back to R in the same Jupyter notebook to finish the work.

I used [tidygeocoder](https://cran.r-project.org/web/packages/tidygeocoder/vignettes/tidygeocoder.html) and [tidycensus](https://cran.r-project.org/web/packages/tidycensus/index.html) to pull in the [American Community Survey](https://www.census.gov/programs-surveys/acs) data easily.

In [15]:
%load_ext rpy2.ipython

The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython


In [20]:
%%R

install.packages("dplyr", repos='http://cran.us.r-project.org', quiet=TRUE)
install.packages("tidygeocoder", repos='http://cran.us.r-project.org', quiet=TRUE)
install.packages("tidycensus", repos='http://cran.us.r-project.org', quiet=TRUE)

library(dplyr)
library(tidygeocoder)
library(tidycensus)
oregon<- read.csv("~/Desktop/output_oregon.csv")

oregon$full <- paste(oregon$address, oregon$city, oregon$state, sep=", ")

library(dplyr)
library(tidygeocoder)
library(tidycensus)

geocoded_df <- oregon %>%
 geocode(address=full,
         method = "census", 
         mode = "single",
         full_results=TRUE,
         api_options = list(census_return_type = 'geographies'),
         quiet=TRUE)

geocoded_df <- 
  geocoded_df %>%
  select(status, lat, long, `geographies.2020 Census Blocks`) %>%
  rename(block = `geographies.2020 Census Blocks`)

block_id <- character(nrow(geocoded_df))

for (i in 1:nrow(geocoded_df)) {
  block.GEOID <- (geocoded_df[i, "block"] %>% unlist)["block.GEOID"] 
  if (is.null(block.GEOID)) {
    block_id[i] <- NA
  } else {
    block_id[i] <- block.GEOID
  }
} 

geocoded_df$block_id <- block_id
geocoded_df <- 
  geocoded_df %>%
  mutate(block_grp_id = substr(block_id, start=1, stop=12),
         tract_id = substr(block_id, start=1, stop=11),
         county_id = substr(block_id, start=1, stop=5)) %>%
  select(status, 
         lat, 
         long, 
         block_grp_id, 
         tract_id, 
         county_id)

### Step 5: Test Differences

Let's use a simple t-test to look at differences between the two groups for household income.

In [21]:
%%R

install.packages("naniar", repos='http://cran.us.r-project.org', quiet=TRUE)
library(naniar)

incomes_by_block_grp <- get_acs(
  geography="block group",
  state="OR",
  variables="B19013_001",
  year=2019,
  geometry=FALSE)


med_incomes_block_grp <-
  left_join(x=geocoded_df, y=incomes_by_block_grp,
            by=c("block_grp_id" = "GEOID")) %>%
  select(status, estimate) %>%
  rename(med_inc_block_grp = estimate)

x = med_incomes_block_grp %>% replace_with_na(replace = list(status = 2))
nonmiss = na.omit(x)
final <- t.test(med_inc_block_grp ~ status, data = nonmiss, var.equal = TRUE)
final

R[write to console]: Getting data from the 2015-2019 5-year ACS

R[write to console]: Using FIPS code '41' for state 'OR'




	Two Sample t-test

data:  med_inc_block_grp by status
t = 2.7734, df = 237, p-value = 0.005989
alternative hypothesis: true difference in means between group 0 and group 1 is not equal to 0
95 percent confidence interval:
  3842.536 22687.299
sample estimates:
mean in group 0 mean in group 1 
       89947.24        76682.32 



### Done! 

I found here that the mean block group household income of the addresses that ordered the test was 89,947 USD compared to 76,682 USD in the group that did not order the test.

This means that in the Portland area - at least - this government intervention more greatly benefited higher income households.